qualang_tools/loops/loops.py

```diff
@@ -0,0 +1,205 @@
+    """Function parametrizing the QUA `for_` loop from a python array.
+
+    :param var: the QUA variable that will be looped over (int or fixed).
+    :param array: a Python list or numpy array containing the values over which `a` will be looping. The spacing must be constant or logarithmic and it must be a Python array.
+    :param var_type: type of the QUA variable 'a', can be either 'int' or 'fixed'.
+    :return: QUA for_ loop parameters (var, init, cond, update) as defined in https://qm-docs.qualang.io/api_references/qua/dsl_main?highlight=for_#qm.qua._dsl.for_.
+    """
+    # Check for array length
+    if len(array) == 0:
+        raise Exception("The array must be of length > 0.")
+    # Check QUA vs python variables
+    if not hasattr(var, "_exp"):
+        raise Exception("The first argument must be a QUA variable.")
+    if hasattr(array[0], "_exp"):
+        raise Exception("The array must be an array of python variables.")
+    # Check array increment
+    if np.isclose(np.std(np.diff(array)), 0):
+        increment = "lin"
+    elif np.isclose(np.std(array[1:] / array[:-1]), 0):
+        increment = "log"
+    else:
+        raise Exception(
+            "The spacing of the input array must be linear or logarithmic. Please use `for_each_()` for arbitrary scans."
+        )
+
+    start = array[0]
+    stop = array[-1]
+    # Linear increment
+    if increment == "lin":
+        step = array[1] - array[0]
+
+        if var_type == "int":
+            # Check that the array is an array of int with integer increments
+            if (
+                not float(step).is_integer()
+                or not float(start).is_integer()
+                or not float(stop).is_integer()
+            ):
+                raise Exception(
+                    "When looping over a QUA int variable, the step and array elements must be integers."
+                )
+            # Generate the loop parameters for positive and negative steps
+            if step > 0:
+                return var, int(start), var <= int(stop), var + int(step)
+            else:
+                return var, int(start), var >= int(stop), var + int(step)
+
+        elif var_type == "fixed":
+            # Check for fixed number overflows
+            if not (-8 <= start < 8) and not (-8 <= stop < 8):
+                raise Exception("fixed numbers are bounded to [-8, 8).")
+
+            # Generate the loop parameters for positive and negative steps
+            if step > 0:
+                return (
+                    var,
+                    float(start),
+                    var < float(stop) + float(step) / 2,
+                    var + float(step),
+                )
+            else:
+                return (
+                    var,
+                    float(start),
+                    var > float(stop) + float(step) / 2,
+                    var + float(step),
+                )
+    # Logarithmic increment
+    elif increment == "log":

```

Comment:
I see.
In any case, if it is not supported it should be noted.
Wait, maybe we can do it using "Cast.mult_int_by_fixed"?
For this we support floats <8 or integers above 8.

Reply:
If I understand well, I think that this multiplication should be made in QUA dsl and not py-qua-tools because in occurs within the QUA for loop, right?
Maybe for now I can just specify that it only works for fixed QUA variables and we'll think of a way of making it work with integer...

Reply:
I'm not sure I follow.
We don't have implicit casting in QUA, which means that:
`assign(a, b*c)` would fail if `a`, `b` and `c` are not of the same type.
This is how it is defined in the DSL, and this we will not change.
Doing these multiplications needs to be explicit using the `Cast` class.
In this module, If someone wants to have a log scale for an integer variable with a fixed variable step, we can easily support it behind th scenes. And this is a common use-case, so I think we should.
We can also support the other way around, (fixed variable, int increase), but that's much less common.

Reply:
Yes, I understand, but there is still a significant rounding issue because `Cast.mul_int_by_fixed()` will convert each point to an integer which will be smaller than the real value and since it is this integer that will be multiplied by the step to get the next one, the error will increase with the number of iterations. A better way would be to keep the fixed point value for the calculation of the next sample and only return the integer.
It is not easy for me to explain like this but for instance, what I am doing now is `item1 = int(item0 * step)`,` item2 = int(item1 * step) = int(int(intem1*step) * step) < int(item1 * step)` and what we need is `item2 = int(item1 * step)`
The bigger the array, the bigger the error...

Reply:
Yes ok I understand, what bothers me is that the QUA array will not match the corresponding python array so it can be confusing for the user if used for calibration purposes...
I can print a warning explaining it, but do you think that `itemn = int(item0 * step**n)` will be possible one day?

Reply:
Manually in a while loop that is easy (as long as step**n is < 8), but in a for loop it is problematic...

Ok. the way I see it there are two cases:
1. In the `from_array` function: Here it is very simple, we can manually check if the values are different or not. If they are equal, no problem. If they are different, this is the "worst" thing, as the user gave an array and got something which is slightly different, so we **definitely** need to output a warning message.
2. In the `logspace` function. This is similar to the above. As comping against `np.logspace` might give a different result... So maybe here we should also output a warning?

Maybe we can also add a function that returns the real array?

Reply:
Yeah, the thing is that the two arrays will almost always be different with logarithmic increment and QUA integers as soon as you have more than 5 points...
I'll put the warnings and create another function to retrieve the exact values taken by the QUA variable